docs/architecture/adr-070-hasher.md

```diff
@@ -0,0 +1,142 @@
+
+Previous discussions such as [#5632](https://github.com/tendermint/tendermint/issues/5631) express interest in using a different hash algorithm for the Merkle tree. [#2186](https://github.com/tendermint/tendermint/issues/2186) and [#2187](https://github.com/tendermint/tendermint/issues/2187) would like to use a faster hash for performance gains. 
+
+In the SDK, [blake3](https://github.com/BLAKE3-team/BLAKE3) is also being considered as a hasher option. Therefore, we can integrate `blake3` if Tendermint and the SDK want to use the same hash package, i.e. `tmhash`.
+
+## Decision
+TBD
+
+## Detailed Design
+To tackle these issues, we can separate into two directions: configurable global hasher, and the custom transaction hash with `hasher` injection.
+
+### Configuable global Hashers
+Some of the components like: `merkle`, `evidence`, `tx`, and `maps` in SDK rely on the `tmhash.Sum()`, we would like to propose using configurable global Hashers instead of it. The `Hashers` provides a container that can store the different hash algorithms. Therefore, developers can assign the `custom hasher` before running the services. 
+
+```go
+package tmhash
+
+type HashType string
+const (
+    SHA256 HashType = "sha256"
+    BLAKE2B HashType = "blake2b"
+    // SHA256_SIMD HashType = "sha256_simd"
+    // other hash type if the Tendermint would like to support by default
+)
+
+// A globle hashers
+var Hashers *HasherContainer
+
+type HasherContainer struct {
+	mainHasher HashType // it indicates which hasher is mainly using in the tendermint components.
+	hasherMap  map[HashType]hash.Hash
+}
+
+func init() {
+	Hashers = initContainer()
+}
+
+func initContainer() *HasherContainer {
+	hc := &HasherContainer{
+		hasherMap: map[HashType]hash.Hash{}}
+
+    // during the hashers initialization, it will adds all hasher have been integreted in 
+    // Tendermint crypto library by default. And the mainHasher will be sha256 by default.    
+	hc.RegisterHasher(SHA256, sha256.New(), true)
+	hc.RegisterHasher(BLAKE2B, blake2b.New(), false)
+
+	return hc
+}
+
+// develop can register the custom hasher into the hashers and set it to the main uses.
+func (hc *HasherContainer) RegisterHasher(ht HashType, hasher hash.Hash, isMain bool) {
+	hc.hasherMap[ht] = hasher
+	if isMain {
+		hc.mainHasher = ht
+	}
+}
```

Comment:
I think we should add `Seal` method. After it's called, we should not be able to register more hashes. This way we protect the runtime for changing the hashes. Otherwise, someone in the runtime can overwrite a hasher.